src/trace/statsd.zig

```diff
@@ -0,0 +1,574 @@
+    @memset(&events_metric_emitted, null);
+
+    var events_timing: [std.meta.fieldNames(EventTiming).len]?EventTimingAggregate = undefined;
+    for (&events_timing, std.meta.fields(EventTiming)) |*event_timing, EventTimingInner| {
+        event_timing.* = .{
+            .event = @unionInit(
+                EventTiming,
+                EventTimingInner.name,
+                struct_longest(EventTimingInner.type),
+            ),
+            .values = .{
+                .duration_min_us = std.math.maxInt(EventTimingAggregate.ValueType),
+                .duration_max_us = std.math.maxInt(EventTimingAggregate.ValueType),
+                .duration_sum_us = std.math.maxInt(EventTimingAggregate.ValueType),
+                .count = std.math.maxInt(EventTimingAggregate.ValueType),
+            },
+        };
+    }
+
+    var iterator = Iterator{
+        .metrics = .{
+            .events_metric = &events_metric,
+            .events_metric_emitted = &events_metric_emitted,
+            .buffer = &buffer,
+        },
+        .timings = .{
+            .events_timing = &events_timing,
+            .buffer = &buffer,
+        },
+    };
+
+    var len_max: u32 = 0;
+    while (iterator.next()) |value| {
+        // Getting a .none here means that either an Aggregate is null (impossible since they are
+        // set above) or, that the maximum buffer length of a single packet has been exceeded.
+        assert(value == .some);
+
+        len_max = @max(value.some.len, len_max);
+    }
+
+    break :blk len_max;
+};
+
+const messages_per_packet = @divFloor(packet_size_max, statsd_line_size_max);
+
+comptime {
+    assert(statsd_line_size_max <= packet_size_max);
+    assert(messages_per_packet > 0);
+}
+
+/// This implementation emits on an open-loop: on the emit interval, it fires off up to
+/// max_packet_count UDP packets, without waiting for completions.
+///
+/// The emit interval needs to be large enough that the kernel will have finished processing them
+/// before emitting again. If not, an error will be logged.
+const max_packet_count = stdx.div_ceil(
+    EventMetric.stack_count + EventTiming.stack_count,
+    messages_per_packet,
+);
+
+const BufferCompletion = struct {
+    buffer: [packet_size_max]u8,
+    completion: IO.Completion = undefined,
+};
+
+const BufferCompletionRing = RingBufferType(*BufferCompletion, .{ .array = max_packet_count });
+
+pub const StatsD = struct {
+    implementation: union(enum) {
+        udp: struct {
+            socket: std.posix.socket_t,
+            io: *IO,
+            send_callback_error_count: u64 = 0,
+        },
+        log,
+    },
+
+    buffer_completions: BufferCompletionRing,
+    buffer_completions_buffer: []BufferCompletion,
+
+    events_metric_emitted: []?EventMetricAggregate,
+
+    /// Creates a statsd instance, which will send UDP packets via the IO instance provided.
+    pub fn init(allocator: std.mem.Allocator, io: *IO, address: std.net.Address) !StatsD {
```

Comment:
```suggestion
    pub fn init_udp(allocator: std.mem.Allocator, io: *IO, address: std.net.Address) !StatsD {
```

(To mirror `init_log()`.)